gh-114785: Remove `Porting from Python2` how-to (#114805)
python/cpython: Doc/howto/pyporting.rst

:author: Brett Cannon

Python 2 reached its official end-of-life at the start of 2020. This means
that no new bug reports, fixes, or changes will be made to Python 2 - it's
no longer supported: see :pep:`373` and
`status of Python versions <https://devguide.python.org/versions>`_.

If you are looking to port an extension module instead of pure Python code,
please see :ref:`cporting-howto`.

The archived python-porting_ mailing list may contain some useful guidance.

Since Python 3.13 the original porting guide was discontinued.
You can find the old guide in the
`archive <https://docs.python.org/3.12/howto/pyporting.html>`_.

Third-party guides
==================

There are also multiple third-party guides that might be useful:

- `Guide by Fedora <https://portingguide.readthedocs.io>`_
- `PyCon 2020 tutorial <https://www.youtube.com/watch?v=JgIgEjASOlk>`_

> - `Guide by Digital Ocean <https://www.digitalocean.com/community/tutorials/how-to-port-python-2-code-to-python-3>`_

Reviewer: Sorry, this was my typo:
```suggestion
- `Guide by DigitalOcean <https://www.digitalocean.com/community/tutorials/how-to-port-python-2-code-to-python-3>`_
```

>    Python 2 reached its official end-of-life at the start of 2020. This means

Reviewer: https://cpython-previews--114805.org.readthedocs.build/en/114805/howto/pyporting.html

We have an abstract but nothing else:

<img width="846" alt="image" src="https://github.com/python/cpython/assets/1324225/5c50ba3c-1901-4b10-b167-85b5a02238b0">

Let's remove the `.. topic:: Abstract` just before this, and dedent this section.

   that no new bug reports, fixes, or changes will be made to Python 2 - it's
   no longer supported: see :pep:`373` and
   `status of Python versions <https://devguide.python.org/versions>`_.

> 
>    This guide is intended to provide you with a path to Python 3 for your
>    code, that includes compatibility with Python 2 as a first step.

Reviewer: This can go:
```suggestion
```

if it's being run under Python 2, then you can run the tool a second time as if
your code is running under Python 3. By running a static type checker twice like
this you can discover if you're e.g. misusing binary data type in one version
of Python compared to another. If you add optional type hints to your code you
can also explicitly state whether your APIs use textual or binary data, helping
to make sure everything functions as expected in both versions of Python.

.. _caniusepython3: https://pypi.org/project/caniusepython3
.. _cheat sheet: https://python-future.org/compatible_idioms.html
.. _coverage.py: https://pypi.org/project/coverage
.. _Futurize: https://python-future.org/automatic_conversion.html
.. _importlib2: https://pypi.org/project/importlib2
.. _Modernize: https://python-modernize.readthedocs.io/
.. _mypy: https://mypy-lang.org/
.. _Porting to Python 3: http://python3porting.com/
.. _Pylint: https://pypi.org/project/pylint

.. _Python 3 Q & A: https://ncoghlan-devs-python-notes.readthedocs.io/en/latest/python3/questions_and_answers.html

.. _pytype: https://github.com/google/pytype
.. _python-future: https://python-future.org/

> .. _python-porting: https://mail.python.org/pipermail/python-porting/

Reviewer: This warning:
``` 
/home/runner/work/cpython/cpython/Doc/howto/pyporting.rst:24: WARNING: Unknown target name: "python-porting".
``` 
is because this target is missing. Let's keep it.

   The archived python-porting_ mailing list may contain some useful guidance.

>    Since Python 3.13 the original official pyporting guide was discontinued.
>    You can still find the old guide in the
>    `archive <https://docs.python.org/3.12/howto/pyporting.html>`_.

Reviewer: ```suggestion
   Since Python 3.13 the original porting guide was discontinued.
   You can find the old guide in the
   `archive <https://docs.python.org/3.12/howto/pyporting.html>`_.
```

Third-party guides
==================

> There are also multiple third-party guides that might be useful:

Reviewer: ```suggestion
There are also multiple third-party guides that might be useful:

```

Newline here to fix formatting?

<img width="814" alt="image" src="https://github.com/python/cpython/assets/1324225/56bee55a-ed20-493d-93e4-9a75a01f9ebb">
